xds/internal/resolver/cluster_specifier_plugin_test.go

```diff
@@ -20,349 +20,466 @@ package resolver
+	rState = val.(resolver.State)
 	if err := rState.ServiceConfig.Err; err != nil {
-		t.Fatalf("ClientConn.UpdateState received error in service config: %v", rState.ServiceConfig.Err)
-	}
-	wantJSON := `{"loadBalancingConfig":[{
-    "xds_cluster_manager_experimental":{
-      "children":{
-        "cluster_specifier_plugin:cspA":{
-          "childPolicy":[{"csp_experimental":{"arbitrary_field":"anything"}}]
-        }
-      }
-    }}]}`
-
-	wantSCParsed := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(wantJSON)
-	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed.Config) {
-		t.Errorf("ClientConn.UpdateState received different service config")
-		t.Error("got: ", cmp.Diff(nil, rState.ServiceConfig.Config))
-		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed.Config))
-	}
-
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes:  []*xdsresource.Route{{Prefix: newStringP(""), ClusterSpecifierPlugin: "cspA"}},
-			},
-		},
-		// Top level csp config here - the value of cspA should get directly
-		// placed as a child policy of xds cluster manager.
-		ClusterSpecifierPlugins: map[string]clusterspecifier.BalancerConfig{"cspA": []map[string]interface{}{{"csp_experimental": cspConfig{ArbitraryField: "changed"}}}},
-	}, nil)
-
-	gotState, err = tcc.stateCh.Receive(ctx)
-	if err != nil {
-		t.Fatalf("Error waiting for UpdateState to be called: %v", err)
+		t.Fatalf("Received error in service config: %v", rState.ServiceConfig.Err)
 	}
-	rState = gotState.(resolver.State)
-	if err := rState.ServiceConfig.Err; err != nil {
-		t.Fatalf("ClientConn.UpdateState received error in service config: %v", rState.ServiceConfig.Err)
-	}
-	wantJSON = `{"loadBalancingConfig":[{
-    "xds_cluster_manager_experimental":{
-      "children":{
-        "cluster_specifier_plugin:cspA":{
-          "childPolicy":[{"csp_experimental":{"arbitrary_field":"changed"}}]
-        }
-      }
-    }}]}`
-
-	wantSCParsed = internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(wantJSON)
+	wantSCParsed = internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(`
+{
+	"loadBalancingConfig": [
+		{
+		  "xds_cluster_manager_experimental": {
+			"children": {
+			  "cluster_specifier_plugin:cspA": {
+				"childPolicy": [
+				  {
+					"csp_experimental": {
+					  "arbitrary_field": "changed"
+					}
+				  }
+				]
+			  }
+			}
+		  }
+		}
+	  ]
+}`)
 	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed.Config) {
-		t.Errorf("ClientConn.UpdateState received different service config")
-		t.Error("got: ", cmp.Diff(nil, rState.ServiceConfig.Config))
-		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed.Config))
+		t.Fatalf("Got service config:\n%s \nWant service config:\n%s", cmp.Diff(nil, rState.ServiceConfig.Config), cmp.Diff(nil, wantSCParsed.Config))
```

Comment:
I know I had this in my test, but is cmp.Diff the cleanest way to do this? Perhaps it is just wanted to know your thoughts.

Reply:
I don't remember the history of this, but I think we decided to use this pattern of multiple `cmp.Diff()` calls (each against `nil`), because it produced better output than a single `cmp.Diff(got, want)` in cases involving service config. We don't use this pattern for anything else. So, clearly someone who wrote it the first time saw this pattern to be more useful in the test error log than the other one. But, the next time anyone of us is working on a test involving service config, we could experiment with `cmp.Diff(got, want)`  and see why we decided to go with the other option.